core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/async/SplittingTransformer.java

```diff
@@ -0,0 +1,335 @@
+        this.returnFuture = Validate.paramNotNull(returnFuture, "returnFuture");
+        this.maximumBufferSize = NumericUtils.saturatedCast(bufferSize);
+        this.maxElements = maxElements;
+        this.isBounded = maxElements != Long.MAX_VALUE;
+    }
+
+    /**
+     * Creates an unbounded publisher.
+     */
+    public SplittingTransformer(AsyncResponseTransformer<ResponseT, ResultT> upstreamResponseTransformer,
+                                long bufferSize, CompletableFuture<ResultT> returnFuture) {
+        this.upstreamResponseTransformer = Validate.paramNotNull(upstreamResponseTransformer, "asyncRequestBody");
+        this.returnFuture = Validate.paramNotNull(returnFuture, "returnFuture");
+        this.maximumBufferSize = NumericUtils.saturatedCast(bufferSize);
+        // unbounded by default
+        this.maxElements = Long.MAX_VALUE;
+        this.isBounded = false;
+    }
+
+    /**
+     * @param downstreamSubscriber the {@link Subscriber} to the individual AsyncResponseTransformer
+     */
+    @Override
+    public void subscribe(Subscriber<? super AsyncResponseTransformer<ResponseT, ResponseT>> downstreamSubscriber) {
+        if (downstreamSubscriber == null) {
+            // §1.09 Subscribe throw NPE on null subscriber
+            throw new NullPointerException("downstreamSubscriber must not be null");
+        }
+        this.downstreamSubscriber = downstreamSubscriber;
+        downstreamSubscriber.onSubscribe(new DownstreamSubscription());
+        if (maxElements < 0) {
+            // §1.04 (optional) must signal on error when fails
+            // §1.09 may reject calls to subscribe if publisher is unable or unwilling to serve them
+            // Rejection must trigger on error after onSubscribe
+            downstreamSubscriber.onError(new IllegalArgumentException("Maximum element to request must be positive"));
+        }
+    }
+
+    /**
+     * The subscription implementation for the subscriber to this SplittingTransformer.
+     */
+    private class DownstreamSubscription implements Subscription {
+
+        @Override
+        public void request(long n) {
+            if (n <= 0) {
+                // §3.09 request zero must signal IllegalArgumentException
+                downstreamSubscriber.onError(new IllegalArgumentException("Amount requested must be positive"));
+                return;
+            }
+            if (isBounded && totalEmitted.get() >= maxElements) {
+                terminate();
+                return;
+            }
+            if (!isCancelled.get()) {
+                long newDemand = outstandingDemand.updateAndGet(current -> {
+                    if (Long.MAX_VALUE - current < n) {
+                        return Long.MAX_VALUE;
+                    }
+                    return current + n;
+                });
+                log.trace(() -> String.format("new outstanding demand: %s", newDemand));
+                if (newDemand > 0) {
```

Comment:
This check seems unnecessary since existing demand should always be non-negative and we have a new demand here.